### PR TITLE
Added options for linking to libraries that are present on system

### PR DESCRIPTION
**Pull request by jzuern**
Added options for linking dynamically to libraries that are present on local system, instead of downloading them (such as libpng, libjpeg, zlib, ...).
This reduces the file size of the shared library libtensorflow.so and adds some security since the libraries such as libpng, libjpeg may be updated independently, without having to rebuild them as part of a tensorflow build from source.

### CODE REVIEW

**Comment by jzuern** [comment]
I signed it!

**Comment by jart** [comment]
@martinwicke Do we know who the best person would be to review a nontrivial cmake build change? If I remember correctly, our friends at Microsoft are big stakeholders in this build?

**Comment by martinwicke** [comment]
By default, @mrry could take a look.

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 1]
```diff
@@ -29,9 +29,22 @@ option(tensorflow_BUILD_ALL_KERNELS "Build all OpKernels" ON)
 option(tensorflow_BUILD_CONTRIB_KERNELS "Build OpKernels from tensorflow/contrib/..." ON)
 option(tensorflow_BUILD_CC_TESTS "Build cc unit tests " OFF)
 option(tensorflow_BUILD_PYTHON_TESTS "Build python unit tests " OFF)
-option(tensorflow_BUILD_SHARED_LIB "Build TensorFlow as a shared library" OFF)
-option(tensorflow_OPTIMIZE_FOR_NATIVE_ARCH "Enable compiler optimizations for the native processor architecture (if available)" ON)
+option(tensorflow_BUILD_SHARED_LIB "Build TensorFlow as a shared library" ON)
```
Comment (rmlarsen) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Why change the default here?

I would think that we should leave all the defaults to reflect the previous behavior.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 2]
```diff
@@ -108,67 +121,231 @@ if (tensorflow_ENABLE_JEMALLOC_SUPPORT)
...
+
+  message(STATUS "do not use system libgif")
+
+  include(gif)
+  list(APPEND tensorflow_EXTERNAL_LIBRARIES ${gif_STATIC_LIBRARIES})
+  list(APPEND tensorflow_EXTERNAL_DEPENDENCIES gif_copy_headers_to_destination)
+
+  include_directories(
+    ${gif_INCLUDE_DIR}
+    )
```
Comment (drpngx) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Put everything on the same line.

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 3]
```diff
@@ -108,67 +121,231 @@ if (tensorflow_ENABLE_JEMALLOC_SUPPORT)
...
-include(lmdb)
-include(eigen)
-include(gemmlowp)
-include(jsoncpp)
-include(farmhash)
-include(fft2d)
-include(highwayhash)
-include(protobuf)
+if(tensorflow_USE_SYSTEM_ZLIB)
+
```
Comment (drpngx) [inline tensorflow/contrib/cmake/CMakeLists.txt]: please remove these extra blank lines everywhere to conform with the current code.

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 4]
```diff
@@ -108,67 +121,231 @@ if (tensorflow_ENABLE_JEMALLOC_SUPPORT)
...
+  message(STATUS "use system zlib")
```
Comment (drpngx) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Capitalize message

File: tensorflow/contrib/cmake/CMakeLists.txt [hunk 5]
```diff
@@ -108,67 +121,231 @@ if (tensorflow_ENABLE_JEMALLOC_SUPPORT)
...
+
+  find_package(ZLIB REQUIRED)
+  include_directories(${ZLIB_INCLUDE_DIR})
+  link_libraries (${ZLIB_LIBRARY})
+
+else(tensorflow_USE_SYSTEM_ZLIB)
```
Comment (drpngx) [inline tensorflow/contrib/cmake/CMakeLists.txt]: Basically this is all the same. Can we create a CMake macro or function?

**Review by drpngx — CHANGES_REQUESTED** [review]

**Comment by vrv** [comment]
Ping for @jzuern  !

**Comment by rmlarsen** [comment]
@jzuern when do you think you'll be able to finish this?

**Comment by jhseu** [comment]
Hi @jzuern, this pull request hasn't been updated in a while. Closing it out as per our policy. Feel free to reopen if you'd like to update the change.

**Pull request closed without merging**